keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// This policy can not be specified more than one.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some Indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// +optional
+	OnSucceededPods *SuccessPolicyOnSucceededPods
+}
+
+// SuccessPolicyOnSucceededPods describes a Job can be succeeded based on succeeded pods.
+type SuccessPolicyOnSucceededPods struct {
+	// Specifies a set of required indexes when .spec.completionMode =
+	// "Indexed" in a text format.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	Indexes string
+
+	// Specifies the required number of indexes when .spec.completionMode =
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or a percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up.
+	// 
+	// +optional
+	Count intstr.IntOrString
+
+	// Specifies the action to be taken on the not finished (succeeded or failed) pods 
+	// when the job achieved the requirements.
+	// Possible values are:
+	//
+	// - Continue: indicates that all pods wouldn't be terminated.
+	// - Terminate: indicates that not finished pods would be terminated.
+	WhenTargetAchieved WhenTargetAchievedSuccessPolicy
+}
+
+// WhenTargetAchievedSuccessPolicy specifies the action to be taken on the pods
+// when the job achieved the requirements.
+// +enum
+type WhenTargetAchievedSuccessPolicy string
+
+const (
+	// All pods wouldn't be terminated when the job reached successPolicy.
+	ContinueWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "Continue"
+
+	// Not finished pods would be terminated when the job reached successPolicy.
+	TerminateWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "Terminate"
+)
+```
+
+Moreover, we validate the following constraints for the `onSucceededPods`:
+- exactly one of the fields `indexes` and `count` is specified
```

Comment:
also check if the format of `indexed` field is correct. Also, if the `count` if expressed as a string has correct "percentage" format. Do we allow point percentages, like "45.5%"?

Reply:
Yes, we do. I will add validations similar to [`ValidateRollingUpdateDaemonSet`](https://github.com/kubernetes/kubernetes/blob/909faa3a9b33161992af0684c5582353b5be6d9b/pkg/apis/apps/validation/validation.go#L387).

Reply:
>  Do we allow point percentages, like "45.5%"?

No, I don't think that we should allow point percentages.
I was supposed to allow only absolute number percentage the same as the deployment's `.spec.strategy.rollingUpdate.maxUnavailable`.

I will add a note to API comments.

Reply:
Done.